src/GovUk.Education.ExploreEducationStatistics.Content.Model/Database/ModelBuilderExtensions.cs

```diff
@@ -0,0 +1,22 @@
+﻿#nullable enable
+using System.Reflection;
+using Microsoft.EntityFrameworkCore;
+
+namespace GovUk.Education.ExploreEducationStatistics.Content.Model.Database;
+
+public static class ModelBuilderExtensions
+{
+    private static readonly MethodInfo? PublicationsFreeTextTableMethodInfo = typeof(ContentDbContext)
+        .GetMethod(nameof(ContentDbContext.PublicationsFreeTextTable),
+            new[] { typeof(string) });
+
+    public static ModelBuilder AddFreeTextTableSupport(this ModelBuilder modelBuilder)
+    {
+        modelBuilder.Entity<FreeTextRank>().HasNoKey().ToTable((string?) null);
+
+        // Map the PublicationsFreeTextTable queryable function to database table-valued function
+        modelBuilder.HasDbFunction(PublicationsFreeTextTableMethodInfo!);
```

Comment:
Some NASA-grade programming tricks here?

Reply:
`PublicationsFreeTextTable` is the signature of a user-defined CLR function. EF Core only cares about the method signature not the body, since it will only be evaluated on the database, not client side: `public IQueryable<FreeTextRank> PublicationsFreeTextTable(string search) =>FromExpression(() => PublicationsFreeTextTable(search));`

`PublicationsFreeTextTableMethodInfo` is the `MethodInfo` used to map that function to the database function I've created with the same name.

```
private static readonly MethodInfo? PublicationsFreeTextTableMethodInfo = typeof(ContentDbContext)
    .GetMethod(nameof(ContentDbContext.PublicationsFreeTextTable),
        new[] { typeof(string) });
```

It's registered with `modelBuilder.HasDbFunction(PublicationsFreeTextTableMethodInfo!);`.

EF core expects to find the database function with the same name. If the name is not the same, or it's in a different schema, or the function is built-in to the database, you can call the methods `.HasName('...')`, `.HasSchema('...')`, `.IsBuiltIn(true)`.

I could have in-lined the configuration without using `MethodInfo` in which case it would have looked like this:

```
modelBuilder.HasDbFunction(typeof(ContentDbContext).GetMethod(nameof(PublicationsFreeTextTable), new[] { typeof(string) })!);
```

There's more info about doing this here, see [User-defined function mapping](https://learn.microsoft.com/en-us/ef/core/querying/user-defined-function-mapping).